src/scheduler/resv_info.c

```diff
@@ -1712,46 +1737,171 @@ confirm_reservation(status *policy, int pbs_sd, resource_resv *unconf_resv, serv
-			if (ret == 1)
-				ret = -2;
-			else
-				ret = -1;
+		if (ninfo != NULL) {
+			if (ninfo->is_down || ninfo->is_offline || ninfo->is_stale || ninfo->is_unknown || ninfo->is_maintenance || ninfo->is_sleeping) {
+				if (ret == 1)
+					ret = -2;
+				else
+					ret = -1;
+			}
+
+			if (resv->resv->resv_queue->running_jobs != NULL)
+				for (j = 0; resv->resv->resv_queue->running_jobs[j] != NULL && ret != -2; j++) {
+					resource_resv *job = resv->resv->resv_queue->running_jobs[j];
+					for (k = 0; job->ninfo_arr[k] != NULL && ret != -2; k++)
+						if (job->ninfo_arr[k]->rank == ninfo->rank) {
+							if (ret == -1)
+								ret = -2;
+							else
+								ret = 1;
+						}
+				}
+
+			if (ns->end_of_chunk)
+				break;
 		}
+	}
 
-		if (resv->resv->resv_queue->running_jobs != NULL)
-			for (j = 0; resv->resv->resv_queue->running_jobs[j] != NULL && ret != -2; j++) {
-				resource_resv *job = resv->resv->resv_queue->running_jobs[j];
-				for (k = 0; job->ninfo_arr[k] != NULL && ret != -2; k++)
-					if (job->ninfo_arr[k]->rank == ninfo->rank) {
-						if (ret == -1)
-							ret = -2;
-						else
-							ret = 1;
-					}
-			}
+	return ret;
+}
+
+/**
+ * @brief "remove" nodes from nspec array by adding them to chunks_to_remove.  The type of node removes
+ * 	is based on node_type.  It can be either an unavailable node or a normal node (or either).
+ * 	If we have 4:A+5:B and num_chunks is 2, we'd remove 2 A's and return.
+ * 
+ * @param[in] resv - reservation to remove nodes from
+ * @param[in] start_of_chk - index into orig_nspec_arr of where to start
+ * @param[in] chk_seq_num - sequence number of chunk to remove nodes from
+ * @param[in] num_chks - number of nodes to remove.  Depending on node_type, it is OK to remove less than this
+ * @param[in, out] chunks_to_remove - chunks from the nspec array to remove
+ * @param[in, out] ctr_ind - index of where to start in chunks_to_remove, and advanced as we add to it
+ * @param[in] node_type - type of node to remove: -1 unavailable, 0 normal, 1 either all with no running jobs
+ * 
+ * @note all nspec chunks to be removed will have their ninfo pointer NULL'd.
+ * 
+ * @return int
+ * @retval number of chunks removed from the nspec array
+ * @retval -1 nspecs are not mapped to select chunks
+ */
+int ralter_remove_nodes(resource_resv *resv, int start_of_chk, int chk_seq_num, int num_chks, int node_type) {
+	int i,k;
+	int cur_chks;
 
-		if (ns->end_of_chunk)
+	cur_chks = num_chks;
+	nspec **nspec_arr;
+
+	nspec_arr = resv->orig_nspec_arr;
+
+	for (i = start_of_chk; nspec_arr[i] != NULL && cur_chks; i++) {
+		int ret;
+		if (nspec_arr[i]->chk == NULL)
+			return -1;
+	
+		if (nspec_arr[i]->chk->seq_num != chk_seq_num)
 			break;
+		ret = check_down_running(resv, i);
+		if (ret == node_type || node_type == 1) {
+			k = i;
+			do {
+				nspec_arr[k]->ninfo = NULL;
+				k++;
+			} while (nspec_arr[k] != NULL && !nspec_arr[k]->end_of_chunk);
+
+			if (nspec_arr[i]->end_of_chunk)
```

Comment:
This a little confusing. Shouldn't we be checking nspec_arr[k]->end_of_chunk instead? Also, if k > i then shouldn't i become k for the next iteration of this loop?